### DISCUSSION TITLE
Integration testing with supertest in nodejs (timeout and cannot log after tests are done)

### DISCUSSION THREAD

**Discussion by G-Salviati** [prisma/prisma#15903] [Q&A]
Hello,
I'm trying to test an endpoint using supertest in nodejs.
I got two problems:

-  "Cannot log after tests are done. Did you forget to wait for something async in your test? Attempted to log "Listening on port: 3000"." 

##### organization_tests.js
```
const request = require("supertest");
const router = require("../../../server/routes/routes");
const { execSync } = require("child_process");
const prisma = require("../../../utils/prisma_client");

beforeAll(async () => {
  await prisma.$connect();
  execSync("npx prisma db push");
  console.log("___________________________________________________________________________________BEFOREALL");
});

afterAll(async () => {
  await prisma.$disconnect();
  console.log("___________________________________________________________________________________AFTERALL");
});

describe("Organization API tests", function () {
  beforeEach(async () => {
    console.log("___________________________________________________________________________________BEFOREEACH");

    const createdData1 = await prisma.organization.create({
      data: {
        apolloId: "apolloId1",
        name: "org1",
        websiteUrl: "https://www.org1.com/",
        linkedinUrl: "https://www.linkedin.com/company/org1/mycompany/verification/",
        primaryDomain: "org1.com",
      },
    });

    console.log(createdData1);
    return createdData1;
  });

  afterEach(async () => {
    console.log("___________________________________________________________________________________AFTEREACH");
    //await prisma.organization.deleteMany({});
  });

  it("GET /organizations/:organizationId", async () => {
    expect(true).toEqual(true);
    /*
    console.log("_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-");
    const organizationId = 1;
    const response = await request(router).get(`/organizations/${organizationId}`).set("Accept", "application/json");

    await expect(response.headers["Content-Type"]).toMatch(/json/);
    await expect(response.status).toEqual(200);
    await expect(response.body.name).toEqual("org1");
    */
  });
});

```

##### routes.js
```
const { app, router, http_logger } = require("../../framework/middlewares");
const PORT = process.env.PORT;

const { setSearchRoutes } = require("./search/searchRoutes");
const { setContactRoutes } = require("./contact/contactRoutes");
const { setOrganizationRoutes } = require("./organization/organizationRoutes");
const { setUpFormats, /*validate,*/ validator, ValidationError } = require("../../utils/validator");

setUpFormats();

setSearchRoutes(router, validator.validate);
setContactRoutes(router, validator.validate);
setOrganizationRoutes(router, validator.validate);

app.use("/searchmgmt", router);

router.use((err, req, res, next) => {
  // Check the error is a validation error
  if (err instanceof ValidationError) {
    // Handle the error
    res.status(400).send(err.validationErrors);
    next();
  } else {
    // Pass error on if not a validation error
    next(err);
  }
});

app.listen(PORT, () => console.log(`Listening on port: ${PORT}`));

module.exports = router;

```

For this problem the test passes with the `expect(true).toEqual(true)` as the only assertion but I get that attempt to console.log after test has finishedand I can't understand why. That's why I pasted the routes.js file here in which the console.log should be executed. The second problem instead is kinda annoying and it prints two errors.

- "TypeError: Cannot read properties of undefined (reading 'apply')
      at Immediate.<anonymous> (node_modules/express/lib/router/index.js:646:15)"
- " thrown: "Exceeded timeout of 5000 ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."" and it points at the 
`it("GET /organizations/:organizationId", async () => {` row

Of course for this second case consider the commented piece of code in the "it" block and remove the `expect(true).toEqual(true)`

NOTES
 - running the test in this case logs every console.log expect for the it block one.
 - the `execSync("npx prisma db push");` in the beforeAll block seems to work because i checked the db tables after running the test and they exist
 - the `await prisma.organization.create` doesn't work. The db table is empty after the test (running it with the `await prisma.organization.deleteMany({});` commented)
 - The db is a docker mssql
 - In routes.js `app.use("/searchmgmt", router);` is used to set the base path. Every call is handled by the router and not by the app

Anyone has any idea? 
Thank you in advance!

**Comment by nurul3101** [comment]
Hello @G-Salviati 👋 

Thank you for providing detailed instructions about the issue that you are facing.

Can you please share the Organization model details so that I can try to reproduce this locally? It would be even better if your repository is open source and I could directly clone it.

You mentioned that `await prisma.organization.create` doesn't work and the db table is empty, is it necessary to make actual database calls? Have you considered [Mocking PrismaClient](https://www.prisma.io/docs/guides/testing/integration-testing#prerequisites)?

**Reply by G-Salviati** [reply]
I'm so sorry but the repo is private and it isn't under my control.
For the second question: yes I would like to make db calls because I've already implemented unit tests in which I mock the prisma client with the help of jest-mock-extended. In this case I want a full fledged integration test with a container running mssql. 
I tried for at least two days to solve this problem but unluckily without any success. 
Another try I will make in a while is to migrate and seed the db via CLI and then try to run tests without the db.create and check if the timeout error happens again. I will reply here.

I'm posting missing code down here:


#### organization model
```
model Organization {
  id            Int       @id @default(autoincrement())
  apolloId      String    @unique
  name          String
  websiteUrl    String?
  linkedinUrl   String?
  primaryDomain String
  industry      String?
  country       String?
  state         String?
  city          String?
  searches      Search[]
  contacts      Contact[]
}
```

#### organizationRoutes.js
```
const OrganizationController = require("../../controllers/organizationController");
const { organizationIdParam, searchOrganizationRequestQuery, saveOrganizationRequestBody } = require("./organization_schemas");

exports.setOrganizationRoutes = function (router, validate) {
  router.get("/organization/:organizationId", validate({ params: organizationIdParam }), OrganizationController.getOrganizationByID);
  router.post("/organizations/search", validate({ query: searchOrganizationRequestQuery }), OrganizationController.searchOrganization);
  router.put("/organizations", validate({ body: saveOrganizationRequestBody }), OrganizationController.saveOrganization);
};

```

#### organizationController.js
```
const OrganizationService = require("../services/organizationService");

exports.getOrganizationByID = async function (req, res) {
  const inputOrganizationId = req.params.organizationId;
  const organizationId = parseInt(inputOrganizationId);

  try {
    const result = await OrganizationService.getOrganizationById(organizationId);

    if (result === null) {
      return res.status(404).end();
    }
    return res.json(result);
  } catch (err) {
    console.error(`getOrganizationById() -> couldn't get the organization: ${err}`);
    return res.status(500).end();
  }
};
```

#### organizationService.js
```
const prisma = require("../../utils/prisma_client");

exports.getOrganizationById = async function (organizationId) {
  const query = {
    where: {
      id: organizationId,
    },
  };
  try {
    const results = await prisma.organization.findUnique(query);
    return results;
  } catch (err) {
    throw err;
  }
};
```

**Reply by G-Salviati** [reply]
Ok I've tried to migrate and seed the db via CLI and run the test commenting the prisma.organization.create in the beforeEach block and nothing changed, the "Cannot read properties of undefined (reading 'apply')" and the timeout errors still there.

**Comment by G-Salviati** [comment]
Maybe I've just discovered what's the main problem.

First of all there are some typo errors but it's not the reason why the db is not feeded with the beforeEach block.

So the typo errors are the following: 
- in organizationRoutes.js for the first route (which is the one tested with supertest) is missing the final 's' of "organizations".
- in organization_tests.js when instead of "request(router).get("organizations/:organizationId ... " the correct way is "request(router).get("/searchmgmt/organizations/:organizationId ... " 

Moreover instead of importing the router it must be imported and used the app as param of the supertest method "request".

----------------------------------------------------------------------------------------------------

Anyway the db problem is not solved and I think due to the unit tests configuration in which i mock the prisma client using the singleton approach. In this way in "jest.config.js" there's this config
```
  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  setupFilesAfterEnv: ["<rootDir>/tests/unit/singleton.js"],
```

and the singleton.js file is the following one
```
const { mockDeep, mockReset } = require("jest-mock-extended");

const prisma = require("../../utils/prisma_client");
const prismaMock = prisma;

jest.mock("../../utils/prisma_client", () => mockDeep());
beforeEach(() => {
  mockReset(prismaMock);
});

module.exports.prismaMock = prisma;

```

So I think that there's some incompatibility between this configuration and the integration testing when the prisma.organization.create tries to feed the real db. 
I tried to comment that configuration line in jest.config.js and delete temporarily the unit test folder (in which there's also the singleton.js file) and everything worked correctly.

Now I ask you if my deduction is correct and how can i solve that keeping the unit tests in the project.

**Comment by G-Salviati** [comment]
Hello @nurul3101,
have you, by any chance, tried to solve this issue? 

Thank you.

**Reply by nurul3101** [reply]
Hello, I haven't tried it yet. There are a lot of moving parts, so it is very difficult to say what could be the issue before I can run it locally. 
Would it be possible for you to create a minimal reproduction similar to your repository, which I can run locally and then I can start debugging it.

**Reply by G-Salviati** [reply]
Hello, I don't know if you've already read the last reply I've done 20 days ago. I think I know what causes the issue but I haven't any idea to solve it. Maybe you can give me some hint based on that. In these days I'll try to make the reproduction you asked for.

**Reply by nurul3101** [reply]
I did go through the response, but for me or anyone in the engineering team to be able to find a solution would require us to run the code locally. Only after that, we can try to make some configuration changes and will be able to verify if unit tests and integration tests coexist and work properly.

**Reply by G-Salviati** [reply]
Ok well, I'll be back in these days with a git repo ready to be tested

**Reply by vladi-stevanovic** [reply]
👋 Hi @G-Salviati I'm checking to see if perhaps you've had time to work on a small repo for us to continue investigating this?

Or, if you've found a solution, we'd love to know how you solved this problem. Feel free to post it in this thread and "Mark as answer" if this discussion is now solved!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
